### ISSUE TITLE
[java] LooseCoupling false-positive with methods only available on concrete collection class

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1218]
Not exactly sure if this is a false-positive, but it is strange.

```java
private final LinkedList<String> list = new LinkedList<>();
...
Iterator<String> descendingIterator = list.descendingIterator();
String value = list.getLast();
```

Technically, the rule is correct. However, the `List` interface doesn't have `descendingIterator`, neither does it have `getLast` and some other useful methods. I guess that when such methods are used, this warning should be disabled? Or do you think that this is bad design nonetheless and people should try to use other methods that are in the interface to simulate those (that certainly is possible, however more inconvenient)? I'm interested on more opinions.

**Comment by jsotuyod** [comment]
Definitely a false positive, although not one easily fixable at the moment. The rule is quite naive and just checks certain collections types, it's not really looking for types that can be relaxed at the moment.

**Comment by jsotuyod** [comment]
Upon a fresh review, this is not a FP. The field could be declared as a `Deque<String>`, with that interface providing all needed methods. When using this approach, PMD doesn't report a violation.

**Comment by boris-petrov** [comment]
@jsotuyod I think you're looking too specifically to this example. I could immediately give you a million others where that doesn't work (there is no interface that has all the needed methods). For example `ensureCapacity` on an `ArrayList`. Of course there are countless others. I believe you should reconsider this but it's up to you of course. :)

**Comment by adangel** [comment]
See also #4615.

If you need to use the implementation details (because you want to use ensureCapacity or whatever), then you should either disable this rule or suppress this warning - because it is not your use case: You deliberately want to use a specific implementation. And the rule simple says: You shouldn't use a specific implementation...
You could also question, why you need to use ensureCapacity... You either have a good reason then and suppressing the warning is ok IMHO, or you don't and then you should maybe just use List.

The rule could be maybe more clear in the description, that it might suggest cases, which would result in compile errors if fixed blindly (so, this rule would be a no-go for autofixes).

Maybe we could add a property for the rule to be more or less strict. However, I like that the rule shows up more often, because then you are  forced to reconsider your code.
